### DISCUSSION TITLE
Could you share more details about kyuubi engine pool mode?

### DISCUSSION THREAD

**Discussion by zhichaoleo** [apache/kyuubi#2555] [Q&A] [answered]
Hi Team,

We want to use kyuubi engine pool mode. But the document is a little simple about this topic. 
As we know, kyuubi engine share mode are: **CONNECTION, USER, GROUP, SERVER**; And the engine pool mode is enabled through _kyuubi.engine.pool.size_. 
From my point of view, I think engine pool mode should work with other engine share mode. Dose it means we can use engine pool mode for any of these engine share mode? 

I tried connection engine share mode and user engine share mode. Seems the pool mode will work under user engine share mode.
![image](https://user-images.githubusercontent.com/13878193/166876849-ee9fbd8b-cc1f-405a-b6f4-1caaa45c07e9.png)

But will not work under connection engine share mode:
![image](https://user-images.githubusercontent.com/13878193/166877711-5ab761fa-c704-4d06-8cef-e22756cddca3.png)

My config:
_# kyuubi engine related
kyuubi.engine.pool.name              kyuubi-engine-pool-alex
kyuubi.engine.pool.size                 3
kyuubi.engine.pool.size.threshold 5_

Could you help to share more details about kyuubi engine pool mode?

Thanks,
Alex

**Comment by pan3793** [comment, marked as answer]
> Dose it means we can use engine pool mode for any of these engine share mode?

Almost right, except to CONNECTION mode.

**Reply by pan3793** [reply]
<img width="1191" alt="image" src="https://user-images.githubusercontent.com/26535726/166882156-c2203540-ce29-487c-9370-4f2b0198cb3b.png">

Figure 1 shows the how the zk-based engine router works.

<img width="1194" alt="image" src="https://user-images.githubusercontent.com/26535726/166881328-2b4a5bc9-a3dc-459f-82fd-7ddcf287c7af.png">

Figure 2 gives an example how user A, B, C, D which belong to group ads to use engine pools. The results are:
- User A will randomly get normal-pool-[1-3]
- User B will randomly get sla-pool-[1-3]
- User C will randomly get sla-pool-[1-2]
- User D will always get sla-pool-3

**Reply by zhichaoleo** [reply]
Got it! Thanks @pan3793 !
